modules/widgets/src/geocoder-widget.tsx

```diff
@@ -0,0 +1,202 @@
+// deck.gl
+// SPDX-License-Identifier: MIT
+// Copyright (c) vis.gl contributors
+
+import {Widget, WidgetProps} from '@deck.gl/core';
+import type {WidgetPlacement, Viewport} from '@deck.gl/core';
+import {FlyToInterpolator, LinearInterpolator} from '@deck.gl/core';
+import {render} from 'preact';
+import {DropdownMenu} from './lib/components/dropdown-menu';
+import {type Geocoder} from './lib/geocode/geocoder';
+import {GeocoderHistory} from './lib/geocode/geocoder-history';
+import {
+  GoogleGeocoder,
+  MapboxGeocoder,
+  OpenCageGeocoder,
+  CoordinatesGeocoder,
+  CurrentLocationGeocoder
+} from './lib/geocode/geocoders';
+
+/** @todo - is the the best we can do? */
+type ViewState = Record<string, unknown>;
+
+const CURRENT_LOCATION = 'current';
+
+/** Properties for the GeocoderWidget */
+export type GeocoderWidgetProps = WidgetProps & {
+  viewId?: string;
+  /** Widget positioning within the view. Default 'top-left'. */
+  placement?: WidgetPlacement;
+  /** Tooltip message */
+  label?: string;
+  transitionDuration?: number;
+  /** Geocoding service selector, for declarative usage */
+  geocoder?: 'google' | 'mapbox' | 'opencage' | 'coordinates' | 'custom';
+  /** Custom geocoding service (Used when geocoder = 'custom') */
+  customGeocoder?: Geocoder;
+  /** API key used for geocoding services */
+  apiKey?: string;
+  /** Whether to use geolocation @note Experimental*/
+  _geolocation?: boolean;
+};
+
+/**
+ * A widget that display a text box that lets user type in a location
+ * and a button that moves the view to that location.
+ * @todo For now only supports coordinates, Could be extended with location service integrations.
+ */
+export class GeocoderWidget extends Widget<GeocoderWidgetProps> {
+  static defaultProps: Required<GeocoderWidgetProps> = {
+    ...Widget.defaultProps,
+    id: 'geolocate',
+    viewId: undefined!,
+    placement: 'top-left',
+    label: 'Geolocate',
+    transitionDuration: 200,
+    geocoder: 'coordinates',
+    customGeocoder: CoordinatesGeocoder,
+    apiKey: '',
+    _geolocation: false
+  };
+
+  className = 'deck-widget-geocoder';
+  placement: WidgetPlacement = 'top-left';
+
+  geocodeHistory = new GeocoderHistory({});
+  addressText: string = '';
```

Comment:
What's the difference between this and the one in the GeocoderHistory class by the same name? Do we need both?